br/pkg/lightning/backend/external/writer.go

```diff
@@ -0,0 +1,353 @@
+	if len(kvs) == 0 {
+		return nil
+	}
+	for _, pair := range kvs {
+		w.batchSize += uint64(len(pair.Key) + len(pair.Val))
+		buf := w.kvBuffer.AllocBytes(len(pair.Key))
+		key := append(buf[:0], pair.Key...)
+		val := w.kvBuffer.AddBytes(pair.Val)
+		w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
+		if w.batchSize >= w.memSizeLimit {
+			if err := w.flushKVs(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// IsSynced implements the backend.EngineWriter interface.
+func (w *Writer) IsSynced() bool {
+	return false
+}
+
+// Close closes the writer.
+func (w *Writer) Close(ctx context.Context) (backend.ChunkFlushStatus, error) {
+	if w.closed {
+		return status(true), nil
+	}
+	w.closed = true
+	defer w.kvBuffer.Destroy()
+	err := w.flushKVs(ctx)
+	if err != nil {
+		return status(false), err
+	}
+
+	logutil.Logger(ctx).Info("close writer",
+		zap.Int("writerID", w.writerID),
+		zap.String("minKey", hex.EncodeToString(w.minKey)),
+		zap.String("maxKey", hex.EncodeToString(w.maxKey)))
+
+	w.writeBatch = nil
+
+	w.onClose(&WriterSummary{
+		WriterID:  w.writerID,
+		Seq:       w.currentSeq,
+		Min:       w.minKey,
+		Max:       w.maxKey,
+		TotalSize: w.totalSize,
+	})
+	return status(true), nil
+}
+
+func (w *Writer) recordMinMax(newMin, newMax tidbkv.Key, size uint64) {
+	if len(w.minKey) == 0 || newMin.Cmp(w.minKey) < 0 {
+		w.minKey = newMin.Clone()
+	}
+	if len(w.maxKey) == 0 || newMax.Cmp(w.maxKey) > 0 {
+		w.maxKey = newMax.Clone()
+	}
+	w.totalSize += size
+}
+
+type status bool
+
+// Flushed implements the backend.ChunkFlushStatus interface.
+func (s status) Flushed() bool {
+	return bool(s)
+}
+
+func (w *Writer) flushKVs(ctx context.Context) (err error) {
+	if len(w.writeBatch) == 0 {
+		return nil
+	}
+
+	dataWriter, statWriter, err := w.createStorageWriter()
+	if err != nil {
+		return err
+	}
+
+	ts := time.Now()
+	var saveBytes uint64
+
+	defer func() {
+		w.currentSeq++
+		err1, err2 := dataWriter.Close(w.ctx), statWriter.Close(w.ctx)
+		if err != nil {
+			return
+		}
+		if err1 != nil {
+			logutil.Logger(ctx).Error("close data writer failed", zap.Error(err))
+			err = err1
+			return
+		}
+		if err2 != nil {
+			logutil.Logger(ctx).Error("close stat writer failed", zap.Error(err))
+			err = err2
+			return
+		}
+		logutil.Logger(ctx).Info("flush kv",
+			zap.Duration("time", time.Since(ts)),
+			zap.Uint64("bytes", saveBytes),
+			zap.Any("rate", float64(saveBytes)/1024.0/1024.0/time.Since(ts).Seconds()))
+	}()
+
+	slices.SortFunc(w.writeBatch[:], func(i, j common.KvPair) int {
+		return bytes.Compare(i.Key, j.Key)
+	})
+
+	w.kvStore, err = NewKeyValueStore(ctx, dataWriter, w.rc, w.writerID, w.currentSeq)
+	if err != nil {
+		return err
+	}
+
+	var kvSize uint64
+	for _, pair := range w.writeBatch {
+		err = w.kvStore.AddKeyValue(pair.Key, pair.Val)
+		if err != nil {
+			return err
+		}
+		kvSize += uint64(len(pair.Key)) + uint64(len(pair.Val))
+	}
+
+	if w.rc.currProp.keys > 0 {
```

Comment:
I prefer we always let `KeyValueStore` maintains the `rangeProperty`. We can add `Close()` to `KeyValueStore` to save the last `rangeProperty`